### PR TITLE
change signed messages comparing logic

### PR DESCRIPTION
**Pull request by egorka-sereda**

### CODE REVIEW

File: integration/qbft/scenarios/integration_test_type.go [hunk 1]
```diff
@@ -61,6 +61,10 @@ type scenarioContext struct {
 	dbs         map[spectypes.OperatorID]basedb.IDb              // 1 per operator, pass same to each instance
 }
 
+type GetterRoot interface {
```
Comment (nkryuchkov) [inline integration/qbft/scenarios/integration_test_type.go]: Please use `spectypes.Root`: https://github.com/bloxapp/ssv-spec/blob/main/types/messages.go#L80

File: integration/qbft/scenarios/integration_test_type.go [hunk 2]
```diff
@@ -558,3 +562,21 @@ func (bn beaconNode) GetAttestationData(slot spec.Slot, committeeIndex spec.Comm
 
 	return data, nil
 }
+
+func validateByRoot(expected, actual GetterRoot) error {
+	expectedRoot, err := expected.GetRoot()
+	if err != nil {
+		return fmt.Errorf("error during geting root from expected: %w", err)
```
Comment (nkryuchkov) [inline integration/qbft/scenarios/integration_test_type.go:569]: Could you please shorten the message?

File: integration/qbft/scenarios/integration_test_type.go [hunk 3]
```diff
@@ -558,3 +562,21 @@ func (bn beaconNode) GetAttestationData(slot spec.Slot, committeeIndex spec.Comm
...
+	}
+
+	actualRoot, err := actual.GetRoot()
+	if err != nil {
+		return fmt.Errorf("error during geting root from actual: %w", err)
```
Comment (nkryuchkov) [inline integration/qbft/scenarios/integration_test_type.go:574]: here as well

File: integration/qbft/scenarios/regular.go [hunk 4]
```diff
@@ -65,7 +62,7 @@ func regularInstanceValidator(consensusData *spectypes.ConsensusData, operatorID
 	return func(actual *protocolstorage.StoredInstance) error {
 		consensusData, err := consensusData.Encode()
 		if err != nil {
-			panic(err)
+			return fmt.Errorf("error during encoding specqbft.ConsensusData: %w", err)
```
Comment (nkryuchkov) [inline integration/qbft/scenarios/regular.go:65]:
```suggestion
			return fmt.Errorf("encode consensus data: %w", err)
```

File: integration/qbft/scenarios/regular.go [hunk 5]
```diff
@@ -74,7 +71,7 @@ func regularInstanceValidator(consensusData *spectypes.ConsensusData, operatorID
 			PrepareJustification:     nil,
 		}).Encode()
 		if err != nil {
-			panic(err)
+			return fmt.Errorf("error during encoding specqbft.ProposalData: %w", err)
```
Comment (nkryuchkov) [inline integration/qbft/scenarios/regular.go:74]: similar to above

File: integration/qbft/scenarios/regular.go [hunk 6]
```diff
@@ -91,53 +88,61 @@ func regularInstanceValidator(consensusData *spectypes.ConsensusData, operatorID
 			panic(err)
 		}
 
-		expectedMsg := spectestingutils.SignQBFTMsg(spectestingutils.Testing4SharesSet().Shares[1], 1, &specqbft.Message{
+		if len(actual.State.ProposeContainer.Msgs[specqbft.FirstRound]) != 1 {
+			return fmt.Errorf("propose countainer expected lenth = 1, actual = %d", len(actual.State.ProposeContainer.Msgs[specqbft.FirstRound]))
```
Comment (nkryuchkov) [inline integration/qbft/scenarios/regular.go:92]: `countainer` -> `container`, `lenth` -> `length`

File: integration/qbft/scenarios/integration_test_type.go (hunk 1 above)
Comment (amir-blox) [inline integration/qbft/scenarios/integration_test_type.go]: Rename to `rootGetter`

File: integration/qbft/scenarios/regular.go [hunk 7]
```diff
@@ -91,53 +88,61 @@ func regularInstanceValidator(consensusData *spectypes.ConsensusData, operatorID
...
 			Data:       proposalData,
 		})
-		if !isMessageExistInRound(expectedMsg, actual.State.ProposeContainer.Msgs[specqbft.FirstRound]) {
-			return fmt.Errorf("poposal message %+v wasn't found at actual.State.ProposeContainer.Msgs[specqbft.FirstRound]", expectedMsg)
+		if err := validateSignedMessage(expectedProposeMsg, actual.State.ProposeContainer.Msgs[specqbft.FirstRound][0]); err != nil { // 0 - means expected always shall be on 0 index
+			return err
 		}
 
-		foundedMsgsCounter := 0 //at the end of test it must be at least 3
+		foundedPreparedMsgsCounter := 0 //at the end of test it must be at least == Quorum
```
Comment (nkryuchkov) [inline integration/qbft/scenarios/regular.go:105]: `founded` -> `found`

**Review by amir-blox — APPROVED** [review]

File: integration/qbft/scenarios/integration_test_type.go [hunk 8]
```diff
@@ -61,6 +61,10 @@ type scenarioContext struct {
 	dbs         map[spectypes.OperatorID]basedb.IDb              // 1 per operator, pass same to each instance
 }
 
+type rootGetter interface {
```
Comment (nkryuchkov) [inline integration/qbft/scenarios/integration_test_type.go:64]: Please use `spectypes.Root`: https://github.com/bloxapp/ssv-spec/blob/main/types/messages.go#L80

File: integration/qbft/tests/all_test.go [hunk 9]
```diff
@@ -1,9 +1,9 @@
 package tests
 
 import (
+	"github.com/bloxapp/ssv-spec/types"
```
Comment (nkryuchkov) [inline integration/qbft/tests/all_test.go:4]: please revert this change

File: integration/qbft/scenarios/regular.go [hunk 10]
```diff
@@ -168,51 +173,36 @@ func regularInstanceValidator(consensusData *spectypes.ConsensusData, operatorID
...
+func validateSignedMessage(expected, actual *specqbft.SignedMessage) error {
+	for i := range expected.Signers {
+		if expected.Signers[i] != actual.Signers[i] {
+			return fmt.Errorf("signers not matching. expected = %+v, actual = %+v", expected.Signers, actual.Signers)
 		}
+	}
 
-		return nil
+	if err := validateByRoot(expected, actual); err != nil &&
+		expected.Signers[0] == actual.Signers[0] {
```
Comment (nkryuchkov) [inline integration/qbft/scenarios/regular.go:196]: Do we need the second check? If `Signers[0]` are different, will the check above return an error?

File: integration/qbft/scenarios/regular.go [hunk 11]
```diff
@@ -168,51 +173,36 @@ func regularInstanceValidator(consensusData *spectypes.ConsensusData, operatorID
...
-			return false
-		}
-
-		actualRoot, err := round[i].GetRoot()
-		if err != nil {
-			return false
-		}
-
-		if bytes.Equal(expectedRoot, actualRoot) {
+		if err := validateSignedMessage(message, round[i]); err == nil {
```
Comment (nkryuchkov) [inline integration/qbft/scenarios/regular.go:205]: Can you please rename `round` to `roundMessages`/`roundMsgs`?

File: integration/qbft/scenarios/regular.go [hunk 12]
```diff
@@ -168,51 +173,36 @@ func regularInstanceValidator(consensusData *spectypes.ConsensusData, operatorID
...
-		}
+		return nil
+	}
+}
 
-		if !bytes.Equal(expectedDecidedMessageRoot, actualDecidedMessageRoot) {
-			return fmt.Errorf("expected and actual decided message roots doesn't matching, expected decided message = %+v, actual = %+v", expectedStoredInstance.DecidedMessage, actual.DecidedMessage)
+func validateSignedMessage(expected, actual *specqbft.SignedMessage) error {
+	for i := range expected.Signers {
+		if expected.Signers[i] != actual.Signers[i] {
```
Comment (nkryuchkov) [inline integration/qbft/scenarios/regular.go:190]: We will also need to check the signature, but I think the signature check has low priority. We can postpone it and add a to-do for now.

File: integration/qbft/tests/all_test.go (hunk 9 above)
Comment (egorka-sereda) [reply integration/qbft/tests/all_test.go:4]: image

Comment (nkryuchkov) [reply integration/qbft/tests/all_test.go:4]: @egorka-sereda This line is needed, but you just moved it from the block with 3rd party imports to the block with standard library imports, which made imports unsorted. Actually, not a big deal, but I'd prefer not to do this change

**Review by nkryuchkov — APPROVED** [review]
